draft-ietf-jsonpath-base.md

```diff
@@ -1515,38 +1515,29 @@ As a `function-argument` in another function expression:
 (2) The arguments of the function expression are well typed when
 each argument of the function can be used for the declared type of the
 corresponding parameter, according to one of the following
-conditions depending on the way the argument is supplied:
-
-{:vspace}
-As a value expressed as a literal:
-: the declared type of the parameter is `ValueType`.
-
-As a singular query:
-: the declared type of the parameter is `ValueType`.
-  In this case:
-
-  * If the query results in a nodelist consisting of a single node, the argument is the value of the node.
-  * If the query results in an empty nodelist, the argument is `Nothing`.
-
-As a general query (`filter-query`, which includes singular query):
-: the declared type of the parameter is `NodesType`.
-
-As a `logical-expr` (which includes `filter-query`):
-: the declared type of the parameter is `LogicalType`.
-  This includes the case that the argument is
-  a query (which includes singular query).
+conditions:
+
+* When the argument is a function expression with declared result type the same as the
+  declared type of the parameter.
```

Comment:
Having this is fine since you don't repeat these cases below.  My table included these cases, so this statement would have been redundant.

Reply:
I was in two minds whether to factor out that statement. There is a wrinkle either way, which was one of the reasons I made this a _draft_ PR. I'd appreciate @cabo's opinion (he'll be tied up with the IETF meeting for some time).

The parameter type major order has the advantage that the parameter types _nearly_ (!) partition the possibilities.

There is one overlap, AFAICT, when a `LogicalType` parameter takes a `LogicalType` function expression argument.  This case is covered both by the first and third conditions (i.e. major bullets). It's not very important because the conclusion is the same from either of those conditions. But it seems inaccurate to introduce the condition list with the words "according to _one of_ the following conditions" (emphasis added), since two of the conditions can apply at the same time in this case.

Reply:
The overlap was avoided by the simplification in https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/commit/0083810df7b04b401afd571633cf98f6dcba0ed1.